zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/appliers/UserTaskAssignmentDeniedApplierTest.java

```diff
@@ -0,0 +1,102 @@
+  private MutableUserTaskState userTaskState;
+
+  /** For setting up the state before testing the applier. */
+  private TestSetup testSetup;
+
+  @BeforeEach
+  public void setup() {
+    userTaskAssignmentDeniedApplierApplier = new UserTaskAssignmentDeniedApplier(processingState);
+    userTaskState = processingState.getUserTaskState();
+    testSetup = new TestSetup(processingState);
+  }
+
+  @Test
+  public void shouldRevertAssigneeIfAssignmentDeniedByTaskListener() {
+    // given
+    final var userTaskKey = 1;
+    final var initialAssignee = "initial";
+    final var newAssignee = "changed";
+
+    final var given = new UserTaskRecord().setUserTaskKey(userTaskKey).setAssignee(initialAssignee);
+
+    testSetup.applyEventToState(userTaskKey, UserTaskIntent.CREATING, given);
+    testSetup.applyEventToState(userTaskKey, UserTaskIntent.CREATED, given);
+    testSetup.applyEventToState(
+        userTaskKey, UserTaskIntent.ASSIGNING, given.setAssignee(newAssignee));
+
+    Assertions.assertThat(userTaskState.getUserTask(userTaskKey).getAssignee())
+        .isEqualTo(initialAssignee);
+    Assertions.assertThat(userTaskState.getLifecycleState(userTaskKey))
+        .isEqualTo(LifecycleState.ASSIGNING);
+
+    // when
+    userTaskAssignmentDeniedApplierApplier.applyState(userTaskKey, given.setAssignee(newAssignee));
+
+    // then
+    Assertions.assertThat(userTaskState.getIntermediateState(userTaskKey))
+        .describedAs("Expect that intermediate state is not present anymore")
+        .isNull();
+    Assertions.assertThat(userTaskState.findRecordRequestMetadata(userTaskKey))
+        .describedAs("Expect that record request metadata is not present anymore")
+        .isEmpty();
+    Assertions.assertThat(userTaskState.getUserTask(userTaskKey).getAssignee())
+        .describedAs("Expect that user task assignee has not been updated")
+        .isEqualTo(initialAssignee);
+    Assertions.assertThat(userTaskState.getLifecycleState(userTaskKey))
+        .describedAs("Expect that lifecycle state is reverted to 'CREATED'")
+        .isEqualTo(LifecycleState.CREATED);
+  }
+
+  private static final class TestSetup {
+
+    private final EventAppliers eventAppliers;
+
+    TestSetup(final MutableProcessingState processingState) {
+      eventAppliers = new EventAppliers();
+      eventAppliers.registerEventAppliers(processingState);
+    }
+
+    /**
+     * Applies the event of the given intent to the state.
+     *
+     * @implNote applies the event using the latest version of the record.
+     * @param intent the intent of the event to apply
+     * @param userTaskRecord data of the event to apply
+     */
+    private void applyEventToState(
+        final long userTaskKey, final UserTaskIntent intent, final UserTaskRecord userTaskRecord) {
+      final int latestVersion = eventAppliers.getLatestVersion(intent);
+      eventAppliers.applyState(userTaskKey, intent, userTaskRecord, latestVersion);
+    }
+  }
```

Comment:
🔧 I noticed that the `TestSetup` private class in this test is a 1:1 duplicate of the one in `UserTaskCorrectedApplierTest`. To reduce duplication and improve reusability, it would be great to extract this class into a shared utility class, perhaps named `ApplierTestSetup` and placing it in a shared `util` package near applier tests. This way, both `UserTaskAssignmentDeniedApplierTest` and `UserTaskCorrectedApplierTest` can reuse it, and it’ll also be available for other applier tests in the future.

Reply:
I have no strong opinion here. I generally prefer small amounts of duplication. I guess this could become some common pattern to use in event applier tests, in that case we could extract it. Personally, I wouldn't spend too much time on it now.